### PR TITLE
Ensure _maxProcessCount is always initialized

### PR DESCRIPTION
**Pull request by liliankasem**
### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (fabiocav) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: I realize this avoids the uninitialized flow, but do we want to do something like we've discussed and put this behind a property to ensure initialization in all cases? 

Also, do we have a race between starting the channels at the web host level and the point where this count is performed?

Comment (liliankasem) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Yup, I'll make this a property.

Good point about a potential race here, any suggestions?

Comment (pragnagopa) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Yes, race conditions are possible here - we have seen some variation of races here earlier as well - see

https://github.com/Azure/azure-functions-host/blob/2597f221d7643b24e206712ab9d8c3a01972b50d/src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs#L405-L421

Instead of relying on the count - refactor the code above -get the count from webhostchannels - this awaits on the Tasks that start in webhostchannelmanager

File: test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs:296]: add check for MaxProcessCount

**Review by pragnagopa — COMMENTED** [review]
added minor comment for the test. Rest looks good.!

File: test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs
Comment (liliankasem) [reply test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs:296]: done!

Comment (liliankasem) [reply test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs:296]: I ended up refactoring this so that we use lazy initialization, and since maxProcessCount is never used outside of this class I made that field private to avoid something else triggering the initialization. So I think we are okay to not check the maxProcessCount field directly here - thoughts?

**Comment by liliankasem** [comment]
@alrod adding you for review as there was some overlap in the code changes we made and want to make sure I'm not breaking your stuff :)

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: move null ref checks on _evironment and _workerconfigs to constructor - if these are null - DI setup is has gone wrong!

File: test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs
Comment (pragnagopa) [reply test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs:296]: exposing internal property to verify in unit tests is OK!

**Review by pragnagopa — APPROVED** [review]
approving with minor comments. Please wait for approval from @fabiocav  as well. Thanks!

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (liliankasem) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: great point, will move!

File: test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs
Comment (liliankasem) [reply test/WebJobs.Script.Tests/Workers/Rpc/RpcFunctionInvocationDispatcherTests.cs:296]: Done, but just FYI I think we were cautious of making it internal so that it didn't get invoked outside of this class i.e giving something else the ability to call `(await _maxProcessCount.Value)` to init the value before we actually need it. technically shouldn't happen as this is not used outside of this class but more of a safety thing

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (fabiocav) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: We should be able to skip `ToArray` here

**Review by fabiocav — APPROVED** [review]
Small comment. Once that is addressed, :shipit:

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs
Comment (liliankasem) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs]: Yup, you're right - I don't know why I had to do that in the first place xD

**Comment by pragnagopa** [comment]
Can you port this to v3.x as well? Thanks!

**Comment by liliankasem** [comment]
> Can you port this to v3.x as well? Thanks!

Can do!

**Comment by liliankasem** [comment]
Alrighty, manual it is :)

**Pull request merged**
